src/SdkCommon/ClientRuntime/ClientRuntime/ExtensibleEnum.cs

```diff
@@ -0,0 +1,121 @@
+        /// <param name="rhs">The second ExtensibleEnum to compare.</param>
+        /// <returns>true if the ExtensibleEnum objects are not equal; false otherwise.</returns>
+        public static bool operator !=(ExtensibleEnum<T> lhs, T rhs) => !object.Equals(lhs, rhs);
+
+        /// <summary>
+        /// Defines explicit conversion from ExtensibleEnum to string.
+        /// </summary>
+        /// <param name="name">ExtensibleEnum to convert.</param>
+        /// <returns>The ExtensibleEnum as a string.</returns>
+        public static explicit operator string(ExtensibleEnum<T> name) => name?.ToString();
+
+        /// <summary>
+        /// Looks up an ExtensibleEnum instance by name, or returns null if the given name does not match one of the
+        /// known values of this ExtensibleEnum.
+        /// </summary>
+        /// <param name="name">Name of the ExtensibleEnum value.</param>
+        /// <returns>An instance of type T with the given name, or null if no such instance could be found.</returns>
+        protected static T Lookup(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(name, "name");
+            }
+
+            T enumValue;
+            return _nameMap.Value.TryGetValue(name, out enumValue) ? enumValue : null;
+        }
+
+        /// <summary>
+        /// Compares the ExtensibleEnum for equality with another ExtensibleEnum.
+        /// </summary>
+        /// <param name="other">The ExtensibleEnum with which to compare.</param>
+        /// <returns>true if the ExtensibleEnum objects are equal; false otherwise.</returns>
+        public bool Equals(T other)
+        {
+            if (object.ReferenceEquals(other, null))
+            {
+                return false;
+            }
+
+            return this._name == other._name;
+        }
+
+        /// <inheritdoc />
+        public override bool Equals(object obj)
+        {
+            return this.Equals(obj as T);
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            return _name.GetHashCode();
+        }
+
+        /// <summary>
+        /// Returns a string representation of the ExtensibleEnum.
+        /// </summary>
+        /// <returns>The ExtensibleEnum as a string.</returns>
+        public override string ToString()
+        {
+            return _name;
+        }
+
+        private static Dictionary<string, T> CreateNameMap()
+        {
+            IEnumerable<FieldInfo> allPublicStaticFields =
+                typeof(T).GetTypeInfo().DeclaredFields.Where(f => f.IsStatic && f.IsPublic);
+
+            IEnumerable<T> allKnownValues =
+                allPublicStaticFields
+                    .Where(f => f.FieldType == typeof(T))
+                    .Select(f => f.GetValue(null))
+                    .OfType<T>();
+
+            return allKnownValues.ToDictionary(v => (string)v, v => v);
```

Comment:
You probably want that to read 
```
allKnownValues.ToDictionary(v => v.ToString(), v => v);
```

The cast will fail if the type isn't explicitly coercible to `string`. `ToString()` won't fail.

Reply:
`T` has a constraint that it must derive from `ExtensibleEnum<T>`, which has an explicit `operator string` implementation. That said, if the intent is to generalize this implementation from `string` to `TVal`, then this should probably be `v => v._value` (or whatever you rename `_name` to).